### PR TITLE
Xray optimizations

### PR DESCRIPTION
**Pull request by salsakran**
Recreating #5976

### CODE REVIEW

**Comment by salsakran** [comment]
@sbelak this is failing an integration test.

**Comment by salsakran** [comment]
overall, this feels _much_ snappier

**Comment by sbelak** [comment]
I've now change the middleware selection to use the pattern @camsaul suggested, added support for Dates and move field-values and query-values from db.metadata-queries as they had no business being there (they were placed there at a time when I didn't understand code layout fully).

File: src/metabase/feature_extraction/feature_extractors.clj [hunk 1]
```diff
@@ -459,29 +458,46 @@
...
+                 :histogram-day     (redux/pre-step
+                                     h/histogram-categorical
+                                     (somef (memfn ^java.util.Date getDay)))
+                 :histogram-month   (redux/pre-step
+                                     h/histogram-categorical
+                                     (somef (memfn ^java.util.Date getMonth)))
+                 :histogram-quarter (redux/pre-step
+                                     h/histogram-categorical
+                                     (somef quarter))}
+                (when-not (or (isa? base_type :type/Date)
```
Comment (senior) [inline src/metabase/feature_extraction/feature_extractors.clj]: Is this not supposed to work with Unix timestamps? Those have a base_type of integer and a special_type of date

**Comment by camsaul** [comment]
where do we set the default value of `true` for `format-rows?`

File: src/metabase/feature_extraction/values.clj [hunk 2]
```diff
@@ -0,0 +1,45 @@
+(ns metabase.feature-extraction.values
+  "Helpers to get data behind various models in a consistent shape."
+  (:require [metabase.db.metadata-queries :as metadata]
+            [metabase.query-processor :as qp]))
+
+(defn field-values
+  "Return all the values of FIELD for QUERY."
+  [field query]
+  (->> (qp/process-query
```
Comment (senior) [inline src/metabase/feature_extraction/values.clj]: Any reason not to use the functions in `metabase.query-processor.middleware.expand` for building this?

File: src/metabase/feature_extraction/values.clj [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+                             query)
+          :middleware {:format-rows? false}})
+       :data
+       :rows
+       (map first)))
+
+(defn query-values
+  "Return all values for QUERY."
+  [db-id query]
+  (-> (qp/process-query
```
Comment (senior) [inline src/metabase/feature_extraction/values.clj]: Same comment as above re: `metabase.query-processor.middleware.expand`

**Comment by senior** [comment]
@camsaul I missed it at first too, it's [here](https://github.com/metabase/metabase/pull/5981/files#diff-8f6fe177b48ed6fb17ab1420a71432d0R23).

**Comment by senior** [comment]
@sbelak if I'm reading it correctly, the toggle for format-rows is in the query map as `{:middleware {:format-rows false}}`? Is the `:middleware` key something that existed before?

File: src/metabase/feature_extraction/feature_extractors.clj (hunk 1 above)
Comment (sbelak) [reply src/metabase/feature_extraction/feature_extractors.clj]: Not entirely sure where, but Unix timestamps already seem to get cast to sql.Timestamp

File: src/metabase/feature_extraction/values.clj (hunk 2 above)
Comment (sbelak) [reply src/metabase/feature_extraction/values.clj]: No (good) reason. Fixed. Thanks.

File: src/metabase/feature_extraction/values.clj (hunk 3 above)
Comment (sbelak) [reply src/metabase/feature_extraction/values.clj]: I don't see where I can use something from expand. Any hints?

**Comment by sbelak** [comment]
I added a top-level :middleware key to query to facilitate toggling of middleware. 

I think it's cleanest if each middleware handles the default case.

**Comment by sbelak** [comment]
There is one thing that's bugging me: sometimes :base_type does not correctly reflect the type returned. For instance metadata for field Orders/created_at in our sample database says :type/DateTime, but what gets returned is java.sql.Date. I've added a dirty workaround, so this is not a blocker and will investigate (ie. bug @senior & @camsaul for help)

File: src/metabase/feature_extraction/feature_extractors.clj [hunk 4]
```diff
@@ -459,29 +463,47 @@
...
+                                     #(when %
+                                        (inc (.getMonth ^java.util.Date %))))
+                 :histogram-quarter (redux/pre-step
+                                     h/histogram-categorical
+                                     (somef quarter))}
+                (when-not (or (isa? base_type :type/Date)
+                              (#{:day :month :year :quarter :week} unit))
+                  {:histogram-hour (redux/pre-step
+                                    h/histogram-categorical
+                                    (somef (memfn ^java.util.Date getHours)))})))
```
Comment (senior) [inline src/metabase/feature_extraction/feature_extractors.clj]: All of these `.getXXXX()` calls on Date are deprecated, I think they want you to use Calendar instead. I don't think it needs to hold up getting merged as I doubt the methods are going anywhere ,but something we should probably fix up.

**Review by senior — APPROVED** [review]

**Pull request merged**